src/operator/control_flow.cc

```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+  };
+  // sanity checks
+  CHECK_EQ(inputs.size() + 2U, (size_t) params.num_args);
+  CHECK_EQ(outputs.size(), (size_t) params.num_outputs);
+  CHECK_EQ(outputs.size(), req.size());
+  CHECK_EQ(inputs.size() + params.num_out_data, outputs.size());
+  for (size_t i = 0; i < (size_t) params.num_out_data; i++)
+    CHECK_EQ(params.max_iterations, outputs[i].shape()[0]);
+  for (const auto &arr : outputs)
+    CHECK_EQ(arr.storage_type(), kDefaultStorage) << "The for operator doesn't support the sparse format";
+  // construct inputs and outputs for cond
+  std::vector<NDArray> cond_inputs, cond_outputs = {NDArray()};
+  WhileLoopState::extract_by_loc(inputs, params.cond_input_locs, &cond_inputs);
+  std::vector<NDArray*> cond_input_ptr, cond_output_ptr;
+  to_ptr_vec(cond_inputs, &cond_input_ptr);
+  to_ptr_vec(cond_outputs, &cond_output_ptr);
+  // a helper function to check is cond is satisfied
+  const auto check_cond = [&cond_input_ptr, &cond_output_ptr, &state]() {
+    // RAII storage
+    struct CPUStorage {
+      uint8_t *data;
+      size_t size;
+      CPUStorage(size_t size): data(new uint8_t[size]), size(size) {
+      }
+      ~CPUStorage() {
+        delete[] data;
+      }
+    };
+    // TODO(Junru): should I reset cond_output_ptr everytime? should I turn off sometime like is_recording?
+    // how to handle the storage?
+    // the SyncCopyToCPU here might be problematic
+    state.cond_op->Forward(nullptr, cond_input_ptr, cond_output_ptr);
+    NDArray *scalar = cond_output_ptr[0];
+    CPUStorage storage(scalar->shape().Size());
+    scalar->SyncCopyToCPU(storage.data, storage.size);
+    return (bool) storage.data[0];
+  };
+  // construct inputs and outputs for func
+  std::vector<NDArray> func_inputs, func_outputs(outputs.size());
+  WhileLoopState::extract_by_loc(inputs, params.func_input_locs, &func_inputs);
+  for (size_t &step = state.n_iterations = 0; step < (size_t) params.max_iterations; ++step) {
+    if (!check_cond()) {
+      break;
+    }
+    // we create func_outputs for the current step:
+    // func_outputs[0: num_out_data] is a slice of outputs[][step]
+    for (size_t i = 0; i < (size_t) params.num_out_data; ++i) {
+      func_outputs[i] = outputs[i].At(step);
+    }
+    // func_outputs[num_out_data: ] are new_loop_vars, need to allocate new memory
+    for (size_t i = params.num_out_data; i < outputs.size(); ++i) {
+      size_t j = i - params.num_out_data;
+      func_outputs[i] = NDArray(inputs[j].shape(), inputs[j].ctx(), true, inputs[j].dtype());
+    }
+    state.Forward(step, func_inputs, req, func_outputs, ctx.need_grad);
+    // func_inputs on the next step:
+    // the output (new_loop_vars) will become the new inputs (loop_vars)
+    for (size_t i = params.num_out_data; i < outputs.size(); ++i) {
+      size_t j = i - params.num_out_data;
+      CHECK_EQ(func_inputs[j].shape(), func_outputs[i].shape());
+      func_inputs[j] = func_outputs[i];
```

Comment:
func_inputs have two types of inputs. Some of them are loop vars and the others are referenced in the Python closure. Do you keep the variables in the closure in the beginning of the vector? The output loop vars have the same order as the input loop vars?

Reply:
Yes, please refer to my comments in _union_inputs